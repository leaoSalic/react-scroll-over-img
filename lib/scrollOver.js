'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.scrollOver = exports.bind = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _environment = require('./environment');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function Scroll() {
    this.scrollEl = false; //要监控的滚动对象
    this.preScrollEl = false; //前元素
    this.scrollData = false; //界面滚动数据
    this.disableHandleCount = 0;
    this.handleList = [];
}

//接口1，设定当前要滚动监控的对象
Scroll.prototype.setElement = function (el) {
    var _this = this;

    var scrollEl = 'object' === (typeof el === 'undefined' ? 'undefined' : _typeof(el)) ? el : document.getElementById(el);
    !scrollEl && function () {
        scrollEl = window.document.body;
    }();
    scrollEl.addEventListener('scroll', function () {
        _this.setScreenScroll();
    });
    this.preScrollEl = this.scrollEl;
    this.scrollEl = scrollEl;
    this.handleList.forEach(function (i) {
        return i.act && i.cb(scrollEl);
    });
};
//获取某个界面元素的位置
Scroll.prototype.getElementTop = function (el) {
    return el.getBoundingClientRect().top + window.document.documentElement.scrollTop + this.scrollEl.scrollTop;
};
Scroll.prototype.setScreenScroll = function () {
    var screenTop = window.pageYOffset ? window.pageYOffset : this.scrollEl.scrollTop,
        screenBottom = screenTop + Number(window.innerHeight ? window.innerHeight : this.scrollEl.clientHeight);
    this.scrollData = {
        top: screenTop,
        bottom: screenBottom
    };
};
Scroll.prototype.addElModifyHandle = function (cb) {
    var id = this.handleList.length;
    this.handleList.push({ act: true, cb: cb });
    return id;
};
Scroll.prototype.removeElModifyHandle = function (id) {
    this.handleList[id].act = false;
    return this;
};

var scroll = new Scroll();
scroll.setElement((0, _environment.get)().scrollDom);
scroll.setScreenScroll();

var getComponentName = function getComponentName(WrappedComponent) {
    return WrappedComponent.displayName || WrappedComponent.name || 'Component';
};

var bind = exports.bind = function bind(dom) {
    scroll.setElement(dom);
};
/**
 * 判断当前元素是否已经滚动到屏幕内，目前只提供一次监控——当元素由下至上滚入屏幕区域时进行事件通知。
 * 需要注意的是组件只允许初始化一次，随后无法通过props修改组件的UI只能内部状态修改。
 * @param inOff 标记当元素滚动进入屏幕区域后，是否移除监听，默认为true
 * @param registerName 注册处理器的名称，默认为 'register'，调用props.registerScrollIn(Dom)，Dom表示真实Dom
 * @param removeName 移除处理器名称，默认为'remove'，调用方式props.removeScrollIn().
 * @param emitName 元素滚入屏幕范围的通知名称，默认为'over'，当元素滚入屏幕时，会被设置为ture,区域外为false。
 * @returns {function(*=)}
 */
var scrollOver = exports.scrollOver = function scrollOver() {
    var inOff = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : true;
    var registerName = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'register';
    var removeName = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 'remove';
    var emitName = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : 'over';

    //扩展变量
    var extParams = [].concat((0, _environment.get)().extParams),
        screen = {};
    extParams.push(registerName);
    extParams.push(removeName);
    extParams.push(emitName);
    screen['extParams'] = extParams;
    return function (Comp) {
        var ScrollOver = function (_React$Component) {
            _inherits(ScrollOver, _React$Component);

            function ScrollOver() {
                var _ref;

                _classCallCheck(this, ScrollOver);

                for (var _len = arguments.length, props = Array(_len), _key = 0; _key < _len; _key++) {
                    props[_key] = arguments[_key];
                }

                var _this2 = _possibleConstructorReturn(this, (_ref = ScrollOver.__proto__ || Object.getPrototypeOf(ScrollOver)).call.apply(_ref, [this].concat(props)));

                _this2.domTop = false; //当前被包装img的顶点位置
                _this2.registerHandle = _this2.registerHandle.bind(_this2);
                _this2.removeHandle = _this2.removeHandle.bind(_this2);
                _this2.elModifyHandle = _this2.elModifyHandle.bind(_this2);
                _this2.checkEmit = _this2.checkEmit.bind(_this2);
                screen[registerName] = _this2.registerHandle;
                screen[removeName] = _this2.removeHandle;
                _this2.state = { over: false };
                return _this2;
            }

            _createClass(ScrollOver, [{
                key: 'elModifyHandle',
                value: function elModifyHandle() {
                    scroll.preScrollEl.removeEventListener('scroll', this.checkEmit);
                    scroll.scrollEl.addEventListener('scroll', this.checkEmit);
                }
            }, {
                key: 'registerHandle',
                value: function registerHandle(dom) {
                    this.domTop = scroll.getElementTop(dom);
                    this.handleId = scroll.addElModifyHandle(this.elModifyHandle);
                    scroll.scrollEl.addEventListener('scroll', this.checkEmit);
                    this.checkEmit();
                }
            }, {
                key: 'removeHandle',
                value: function removeHandle() {
                    scroll.removeElModifyHandle(this.handleId);
                    scroll.scrollEl.removeEventListener('scroll', this.checkEmit);
                }
            }, {
                key: 'checkEmit',
                value: function checkEmit() {
                    var _this3 = this;

                    this.domTop < scroll.scrollData.bottom && function () {
                        inOff && _this3.removeHandle();
                        _this3.setState({ over: true });
                    }();
                }
            }, {
                key: 'componentWillUnmount',
                value: function componentWillUnmount() {
                    scroll.removeElModifyHandle(this.handleId);
                    scroll.scrollEl.removeHandle('scroll', this.checkEmit);
                }
            }, {
                key: 'shouldComponentUpdate',
                value: function shouldComponentUpdate(nextProps, nextState) {
                    return 'all' === (0, _environment.get)().renderMode ? true : this.state !== nextState; //不允许外部props变更导致组件进行融合计算，只允许一次初始化
                }
            }, {
                key: 'render',
                value: function render() {
                    screen[emitName] = this.state.over;
                    var props = Object.assign({}, this.props, screen);
                    return _react2.default.createElement(Comp, props);
                }
            }]);

            return ScrollOver;
        }(_react2.default.Component);

        ScrollOver.displayName = 'scrollOver(' + getComponentName(Comp) + ')';
        return ScrollOver;
    };
};