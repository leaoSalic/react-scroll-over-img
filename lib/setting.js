/**
 * 全局设置入口
 * @param options
 */
'use strict';

var setting = function setting(options) {
    naming.forEach(function (i) {
        var option = options[i];
        option && (defines[i] = option);
    });
},
    getting = function getting() {
    return options;
},
    naming = ['onOff', 'maskStyle'],
    defines = {
    onOff: true, //全局图片延迟加载开关[true|false]。设置为true时图片滚动进入浏览区域后加载，设置为false时候组件一生成就加载
    maskStyle: {
        filter: 'blur(.5rem)'
    }
};

module.exports = setting;
module.exports.default = module.exports;
module.exports.getting = getting;