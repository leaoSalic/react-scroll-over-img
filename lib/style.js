'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.addFilter = undefined;

var _environment = require('./environment');

var options = (0, _environment.get)();

/**
 * 添加自定义的样式
 * @returns {*}
 */
var addFilter = exports.addFilter = function addFilter() {
    var document = 'undefined' !== typeof window && window.document,
        el = document && document.createElement('style');
    if (el && el.append) {
        el.setAttribute('type', 'text/css');
        el.append('.' + options.filterName + '{' + options.filter + '}');
        document.head.appendChild(el);
        return options.filterName;
    } else {
        return false;
    }
};

exports.default = {
    addFilter: addFilter
};