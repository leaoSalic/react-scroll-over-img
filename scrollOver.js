'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _environment = require('./lib/environment');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

//---处理区域滚动
var scrollData = null;
var setScreenScroll = function setScreenScroll() {
    var screenTop = window.pageYOffset ? window.pageYOffset : window.document.documentElement.scrollTop,
        screenBottom = screenTop + Number(window.innerHeight ? window.innerHeight : document.documentElement.clientHeight);
    scrollData = {
        top: screenTop,
        bottom: screenBottom
    };
},


/**
 * 获取当前对象的全局位置高度
 * @param ref
 * @return {number}
 */
getDomTop = function getDomTop(ref) {
    return ref.getBoundingClientRect().top + window.document.documentElement.scrollTop + window.document.body.scrollTop;
};
setScreenScroll();
window.addEventListener('scroll', function () {
    setScreenScroll();
});
//-----


/**
 * 判断当前元素是否已经滚动到屏幕内，目前只提供一次监控——当元素由下至上滚入屏幕区域时进行事件通知。
 * 需要注意的是组件只允许初始化一次，随后无法通过props修改组件的UI只能内部状态修改。
 * @param inOff 标记当元素滚动进入屏幕区域后，是否移除监听，默认为true
 * @param registerName 注册处理器的名称，默认为 'register'，调用props.registerScrollIn(Dom)，Dom表示真实Dom
 * @param removeName 移除处理器名称，默认为'remove'，调用方式props.removeScrollIn().
 * @param emitName 元素滚入屏幕范围的通知名称，默认为'over'，当元素滚入屏幕时，会被设置为ture,区域外为false。
 * @returns {function(*=)}
 */
var scrollOver = function scrollOver() {
    var inOff = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : true;
    var registerName = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'register';
    var removeName = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 'remove';
    var emitName = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : 'over';

    return function (Comp) {
        var ScrollOver = function (_React$Component) {
            _inherits(ScrollOver, _React$Component);

            function ScrollOver() {
                var _ref;

                _classCallCheck(this, ScrollOver);

                for (var _len = arguments.length, props = Array(_len), _key = 0; _key < _len; _key++) {
                    props[_key] = arguments[_key];
                }

                var _this = _possibleConstructorReturn(this, (_ref = ScrollOver.__proto__ || Object.getPrototypeOf(ScrollOver)).call.apply(_ref, [this].concat(props)));

                _this.registerHandle = _this.registerHandle.bind(_this);
                _this.removeHandle = _this.removeHandle.bind(_this);
                _this.scrollHandle = _this.scrollHandle.bind(_this);
                _this.state = { over: false };
                return _this;
            }

            _createClass(ScrollOver, [{
                key: 'registerHandle',
                value: function registerHandle(dom) {
                    this.domTop = getDomTop(dom);
                    window.addEventListener('scroll', this.scrollHandle);
                    this.checkEmit();
                }
            }, {
                key: 'removeHandle',
                value: function removeHandle() {
                    window.removeEventListener('scroll', this.scrollHandle);
                }
            }, {
                key: 'componentWillUnmount',
                value: function componentWillUnmount() {
                    window.removeEventListener('scroll', this.scrollHandle);
                }
            }, {
                key: 'scrollHandle',
                value: function scrollHandle() {
                    this.checkEmit();
                }
            }, {
                key: 'checkEmit',
                value: function checkEmit() {
                    var _this2 = this;

                    this.domTop < scrollData.bottom && function () {
                        inOff && _this2.removeHandle();
                        _this2.setState({ over: true });
                    }();
                }
            }, {
                key: 'shouldComponentUpdate',
                value: function shouldComponentUpdate(nextProps, nextState) {
                    return 'all' === (0, _environment.get)().renderMode ? true : this.state !== nextState; //不允许外部props变更导致组件进行融合计算，只允许一次初始化
                }
            }, {
                key: 'render',
                value: function render() {
                    var screen = {};
                    screen[registerName] = this.registerHandle;
                    screen[removeName] = this.removeHandle;
                    screen[emitName] = this.state.over;
                    var props = Object.assign({}, this.props, screen);
                    return _react2.default.createElement(Comp, props);
                }
            }]);

            return ScrollOver;
        }(_react2.default.Component);

        ScrollOver.displayName = 'scrollOver(' + getComponentName(Comp) + ')';
        return ScrollOver;
    };
};

var getComponentName = function getComponentName(WrappedComponent) {
    return WrappedComponent.displayName || WrappedComponent.name || 'Component';
};

module.exports = scrollOver;
module.exports.default = module.exports;