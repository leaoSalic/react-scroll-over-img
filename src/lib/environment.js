/**
 * 全局设置入口
 * @param options
 */
'use strict';
export const get = () => {
    initValue()
    return defines
}
/**
 * 运行时动态设置options，Img组件相关的defaultProps参数无法在运行时配置[]
 * @param options
 */
const set = (options) => {
    options && Object.keys(defines).forEach(i=>{
        const option = options[i]
        'undefined' !== typeof option && (defines[i] = option)
    })
}
const initValue = () => {
        !hasInit && (()=>{
            try{
                set(__scrollOverImgOptions)
            }catch(error){}
            hasInit = true
        })()
    },
    defines = {
        renderMode: 'all', //渲染模式。['all'|'none'], all表示父组件任何渲染都会导致img直接渲染，none表示父组件发生任何渲染都不会引起内部渲染，默认为all。设置为node可以明显提升渲染效率
        onOff: true, //全局图片延迟加载开关[true|false]。设置为true时图片滚动进入浏览区域后加载，设置为false时候组件一生成就加载
        flowDelay: 10, //图片处理流水线启动时间,ms。
        flowNUmber: 2,//流水线个数
        empty: 'data:image/gif;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVQImWNgYGBgAAAABQABh6FO1AAAAABJRU5ErkJggg==', //空白图片
        filterName: 'react-scroll-over-img-filter', //过滤样式名称
        filter: 'filter: blur(.5rem);filter: progid:DXImageTransform.Microsoft.Blur(PixelRadius=10, MakeShadow=false);', //加载页面的毛玻璃效果
        extParams : ['onOff', 'loadSrc', 'loadClassName', 'extParams'] // Img组件的扩展组件 ，除此之外还有'register', 'remove', 'over'，名称有scrollOver组件内部决定
    }
let hasInit = false