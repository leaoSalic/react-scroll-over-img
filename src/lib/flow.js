'use strict';

/**
 * 流水线加载工具，实现一次只加载一张图片，阻止并发。
 * ---------------------------------------
 * 工具类只提供一个外部接口：
 * registerFLow(src, callback)
 * src：用于指定要加载图片的地址。
 * callback：图片加载完成的回调。加载完成后回执行callback({suc:[boolean],src:[string]})。出现错误时，会在console输出错误信息。
 * ---------------------------------------
 * 全局配置
 * 可以通过 environment.set(options)方法全局配置流水线延迟启动时间和并行加载个数。
 * import env from 'react-scroll-over-img'
 * env.set({
 *     flowDelay: 10,
 *     flowNumber: 2
 * })
 * flowDelay：延迟启动时间，单位毫秒。设置得越长，启动时间越慢，但是会等待更多的图片一次性加入到队列中。默认10Ms
 * flowNumber：流水线，可以同时加载的图片数。
 */

import {get} from './environment'

const environment = get()

/**
 * 图片流水线，在动态渲染时，所有的图片都会添加到流水线上，然后逐一渲染
 * @type {Array}
 */
const flowList = [];
let isStop = true, //处理标记，表示当前流水线是否处于处理中
    timerId; //时间计时器对象

/**
 * (接口)注册一个加载原图回调，向流水线增加一个处理节点
 * @param {string} src 原始图片路径
 * @param {object} img 对象，可以通过new Image创建
 * @param {function} callback 回调函数 (src)=>{}
 */
const registerFLow = (src, img, callback) => {
        //向流水线队尾部增加对象
        flowList.push({src, img, callback});
        isStop && (() => {
            timerId && clearTimeout(timerId);
            timerId = setTimeout(loadOneStream(), environment.flowDelay);
        })()
    },

    /**
     * 运行流水线
     */
    loadOneStream = () => {
        isStop = false;
        const item = flowList.pop();//处理队列头部对象
        item ? loadImg(item) : isStop = true;
    },

    /**
     * 加载完成处理
     * @param cb
     * @param src
     */
    resultHandle = (cb, src) => {
        cb({suc: 'NONE' !== src, src})
        loadOneStream()
    },

    /**
     * 加载单个图片
     * --------------
     * 使用jest单元测试时，Image对象的onerror、onabort、onload句柄均不会执行。
     * 因此会将environment.isUnitTest设置为true用于进行相关操作。
     * 当该参数存在值时候，会直接返回一个img对象。在src参数中
     * --------------
     * @param item
     */
    loadImg = (item) => {
        const {src, img, callback} = item
        img.src = src
        img.onerror = img.onabort = () =>
            resultHandle(callback, 'NONE')
        if (img.complete) {
            resultHandle(callback, src)
        } else {
            img.onload = () =>
                resultHandle(callback, src)
        }
    };

module.exports = registerFLow
module.exports.default = module.exports