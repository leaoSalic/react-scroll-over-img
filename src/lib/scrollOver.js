'use strict';
import React from 'react'
import {get} from './environment'

const __element = get().scrollDom,
    __offsetY = get().scrollOffset

function Scroll() {
    const _this = this
    this.scrollEl = false//要监控的滚动对象
    this.preScrollEl = false//前滚动对象
    this.handleList = []
    this.viewHeight = window ? window.document.body.clientHeight : 0 //当前浏览器的视口高度
    this.resize = function (event) {
        _this.viewHeight = window ? window.document.body.clientHeight : 0
    }
}

Scroll.prototype.setElement = function (el) {
    let scrollEl = 'object' === typeof el ? el : document.getElementById(el)
    !scrollEl && (() => {
        scrollEl = window.document.body
    })()
    this.preScrollEl = this.scrollEl
    this.scrollEl = scrollEl
    this.handleList.forEach(i => i.act && i.cb(scrollEl))
}
/**
 * 判断指定元素是否进入视口（用户可视区域）
 * @param el
 * @returns {boolean}
 */
Scroll.prototype.elementInView = function (el) {
    return __offsetY < this.viewHeight - el.getBoundingClientRect().top
}
/**
 * 设定当滚动元素变更时触发的回调函数
 * @param cb
 * @returns {Number}
 */
Scroll.prototype.addElModifyHandle = function (cb) {
    const id = this.handleList.length
    this.handleList.push({act: true, cb})
    return id
}
/**
 * 移除回调函数
 * @param id
 * @returns {Scroll}
 */
Scroll.prototype.removeElModifyHandle = function (id) {
    this.handleList[id].act = false
    return this
}
/**
 * 注册滚动监听回调
 * @param cb
 */
Scroll.prototype.addListener = function (cb) {
    this.scrollEl && this.scrollEl.addEventListener('scroll', cb)
}
/**
 * 移除滚动监听回调
 * @param cb
 */
Scroll.prototype.removeListener = function (cb) {
    this.scrollEl && this.scrollEl.removeEventListener('scroll', cb)
}
/**
 * 重设滚动监听回调
 * @param cb
 */
Scroll.prototype.setListener = function (cb) {
    this.preScrollEl && this.preScrollEl.removeEventListener('scroll', cb)
    this.scrollEl && this.scrollEl.addEventListener('scroll', cb)
}

const scroll = new Scroll()
window.addEventListener('resize', scroll.resize)
scroll.setElement(__element)
const getComponentName = WrappedComponent => {
    return WrappedComponent.displayName || WrappedComponent.name || 'Component';
}

export const bind = (dom) => {
    scroll.setElement(dom)
}
/**
 * 判断当前元素是否已经滚动到屏幕内，目前只提供一次监控——当元素由下至上滚入屏幕区域时进行事件通知。
 * 需要注意的是组件只允许初始化一次，随后无法通过props修改组件的UI只能内部状态修改。
 * @param inOff 标记当元素滚动进入屏幕区域后，是否移除监听，默认为true
 * @param registerName 注册处理器的名称，默认为 'register'，调用props.registerScrollIn(Dom)，Dom表示真实Dom
 * @param removeName 移除处理器名称，默认为'remove'，调用方式props.removeScrollIn().
 * @param emitName 元素滚入屏幕范围的通知名称，默认为'over'，当元素滚入屏幕时，会被设置为ture,区域外为false。
 * @returns {function(*=)}
 */
export const scrollOver = (inOff = true, registerName = 'register', removeName = 'remove', emitName = 'over') => {
    //扩展变量
    const extParams = [].concat(get().extParams), screen = {}
    extParams.push(registerName)
    extParams.push(removeName)
    extParams.push(emitName)
    screen['extParams'] = extParams
    return (Comp) => {
        class ScrollOver extends React.Component {
            constructor(...props) {
                super(...props)
                this.element = false //监控的img元素
                this.registerHandle = this.registerHandle.bind(this)
                this.removeHandle = this.removeHandle.bind(this)
                this.elModifyHandle = this.elModifyHandle.bind(this)
                this.checkEmit = this.checkEmit.bind(this)
                screen[registerName] = this.registerHandle
                screen[removeName] = this.removeHandle
                this.state = {over: false}
            }

            elModifyHandle() {
                scroll.setListener(this.checkEmit)
            }

            registerHandle(el) {
                this.element = el
                this.handleId = scroll.addElModifyHandle(this.elModifyHandle)
                scroll.addListener(this.checkEmit)
                this.checkEmit();
            }

            removeHandle() {
                scroll.removeElModifyHandle(this.handleId)
                scroll.removeListener(this.checkEmit)
            }

            checkEmit() {
                scroll.elementInView(this.element) && (() => {
                    inOff && this.removeHandle()
                    this.setState({over: true})
                })();
            }

            componentWillUnmount() {
                scroll.removeElModifyHandle(this.handleId)
                scroll.removeListener(this.checkEmit)
            }

            shouldComponentUpdate(nextProps, nextState) {
                return 'all' === get().renderMode ? true : this.state !== nextState//不允许外部props变更导致组件进行融合计算，只允许一次初始化
            }

            render() {
                screen[emitName] = this.state.over
                const props = Object.assign({}, this.props, screen);
                return <Comp {...props} />
            }
        }

        ScrollOver.displayName = `scrollOver(${getComponentName(Comp)})`;
        return ScrollOver;
    }
};