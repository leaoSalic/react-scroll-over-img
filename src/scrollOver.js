'use strict';
import React from 'react'
import {get} from './lib/environment'

//---处理区域滚动
let scrollData = null;
const setScreenScroll = ()=> {
        const screenTop = window.pageYOffset ? window.pageYOffset : window.document.documentElement.scrollTop,
            screenBottom = screenTop + Number(window.innerHeight ? window.innerHeight : document.documentElement.clientHeight);
        scrollData = {
            top: screenTop,
            bottom: screenBottom
        }
    },

    /**
     * 获取当前对象的全局位置高度
     * @param ref
     * @return {number}
     */
    getDomTop = (ref) => {
        return ref.getBoundingClientRect().top + window.document.documentElement.scrollTop + window.document.body.scrollTop;
    };
setScreenScroll();
window.addEventListener('scroll', ()=> {
    setScreenScroll();
});
//-----


/**
 * 判断当前元素是否已经滚动到屏幕内，目前只提供一次监控——当元素由下至上滚入屏幕区域时进行事件通知。
 * 需要注意的是组件只允许初始化一次，随后无法通过props修改组件的UI只能内部状态修改。
 * @param inOff 标记当元素滚动进入屏幕区域后，是否移除监听，默认为true
 * @param registerName 注册处理器的名称，默认为 'register'，调用props.registerScrollIn(Dom)，Dom表示真实Dom
 * @param removeName 移除处理器名称，默认为'remove'，调用方式props.removeScrollIn().
 * @param emitName 元素滚入屏幕范围的通知名称，默认为'over'，当元素滚入屏幕时，会被设置为ture,区域外为false。
 * @returns {function(*=)}
 */
const scrollOver = (inOff = true, registerName = 'register', removeName = 'remove', emitName = 'over') => {
    //扩展变量
    const extParams = [].concat(get().extParams), screen = {}
    extParams.push(registerName)
    extParams.push(removeName)
    extParams.push(emitName)
    screen['extParams'] = extParams
    return (Comp) => {
        class ScrollOver extends React.Component {
            constructor(...props) {
                super(...props)
                this.registerHandle = this.registerHandle.bind(this)
                this.removeHandle = this.removeHandle.bind(this)
                this.scrollHandle = this.scrollHandle.bind(this)
                screen[registerName] = this.registerHandle
                screen[removeName] = this.removeHandle
                this.state = {over: false}
            }

            registerHandle(dom) {
                this.domTop = getDomTop(dom)
                window.addEventListener('scroll', this.scrollHandle)
                this.checkEmit();
            }

            removeHandle() {
                window.removeEventListener('scroll', this.scrollHandle)
            }

            componentWillUnmount() {
                window.removeEventListener('scroll', this.scrollHandle)
            }

            scrollHandle() {
                this.checkEmit()
            }

            checkEmit() {
                this.domTop < scrollData.bottom && (() => {
                    inOff && this.removeHandle()
                    this.setState({over: true})
                })();
            }

            shouldComponentUpdate(nextProps, nextState) {
                return 'all' === get().renderMode ? true : this.state !== nextState//不允许外部props变更导致组件进行融合计算，只允许一次初始化
            }

            render() {
                screen[emitName] = this.state.over
                const props = Object.assign({}, this.props, screen);
                return <Comp {...props} />
            }
        }

        ScrollOver.displayName = `scrollOver(${getComponentName(Comp)})`;
        return ScrollOver;
    }
};

const getComponentName = WrappedComponent => {
    return WrappedComponent.displayName || WrappedComponent.name || 'Component';
}

module.exports = scrollOver
module.exports.default = module.exports