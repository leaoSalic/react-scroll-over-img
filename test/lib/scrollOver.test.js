import React from 'react'
import {scrollOver} from '../../lib/scrollOver'
import testRender from 'react-test-renderer'

test('scrollOver component', ()=>{
    const Wrapper = props =>
        <div>测试数据</div>
    const Comp = scrollOver()(Wrapper)
    const result = testRender.create(<Comp />)
    let tree = result.toJSON()
    expect(tree).toMatchSnapshot();
})