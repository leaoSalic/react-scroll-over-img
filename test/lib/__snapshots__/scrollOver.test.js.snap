// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scrollOver component 1`] = `
<div>
  测试数据
</div>
`;
